src/Features/Core/Diagnostics/AbstractHostDiagnosticUpdateSource.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal abstract class AbstractHostDiagnosticUpdateSource : IDiagnosticUpdateSource
+    {
+        private static ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<DiagnosticData>> _analyzerHostDiagnosticsMap =
```

Comment:
Cleanup is called from Dispose method of VisualStudioAnalyzer, which I believe creates a new AnalyzerFileReference for every provided analyzer reference.